src/codegen/compile.ml

```diff
@@ -10451,36 +10884,36 @@ and compile_prim_invocation (env : E.t) ae p es at =
       get_val ^^
       G.i (Convert (Wasm.Values.I32 I32Op.WrapI64))
     | Int8, Int16 ->
-      SR.Vanilla,
-      compile_exp_vanilla env ae e ^^
+      SR.UnboxedWord32 Int16,
+      compile_exp_as env ae (SR.UnboxedWord32 Int8) e ^^
       compile_shrS_const 8l
     | Int16, Int32 ->
-      SR.Vanilla,
-      compile_exp_vanilla env ae e ^^
-      compile_shrS_const 15l (* resulting Int32 will always be unboxed *)
+      SR.UnboxedWord32 Int32,
+      compile_exp_as env ae (SR.UnboxedWord32 Int16) e ^^
+      compile_shrS_const 16l (* lsb_adjust? *)
     | Int32, Int64 ->
-      SR.UnboxedWord64,
-      compile_exp_as env ae SR.UnboxedWord32 e ^^
+      SR.UnboxedWord64 Int64,
+      compile_exp_as env ae (SR.UnboxedWord32 Int32) e ^^
       G.i (Convert (Wasm.Values.I64 I64Op.ExtendSI32))
     | Int16, (Int8 as pty) ->
-      SR.Vanilla,
+      SR.UnboxedWord32 Int8,
       let num_bits = (TaggedSmallWord.bits_of_type pty) in
       let set_val, get_val = new_local env "convertee" in
-      compile_exp_vanilla env ae e ^^
+      compile_exp_as env ae (SR.UnboxedWord32 Int16)  e ^^
       set_val ^^
       get_val ^^
       compile_shl_const (Int32.of_int num_bits) ^^
       compile_shrS_const (Int32.of_int num_bits) ^^
       get_val ^^
-      compile_eq env Type.(Prim Nat16) ^^
+      compile_eq env Type.(Prim Nat16) ^^ (* TBR *)
```

Comment:
Although the implementation results in the same code, I am not sure whether this should rather be: 
```suggestion
      compile_eq env Type.(Prim Int16) ^^ 
```

Reply:
That was the original code, but seems like a benign change. Done.